src/mscorlib/src/System/String.cs

```diff
@@ -568,12 +568,14 @@ private unsafe static int CompareOrdinalHelper(String strA, String strB)
 #endif
                 
                 DiffNextInt:
-                int order;
-                if ( (order = (int)*a - (int)*b) != 0) {
-                    return order;
-                }
-                Contract.Assert( *(a+1) != *(b+1), "This char must be different if we reach here!");
-                return ((int)*(a+1) - (int)*(b+1));
+                if (*a == *b) goto ReturnDifference;
+
+                DiffOffset1:
+                Contract.Assert(*(a + 1) != *(b + 1), "This char must be different if we reach here!");
+                a += 1; b += 1;
```

Comment:
You're probably right; I had just refactored that to make the code size smaller (so there was only one return point, so multiple of [these](https://gist.github.com/jamesqo/de2fa42f6ffe30e1bbabf3e893051320#file-compareordinalhelper-x86-asm-L157-L163) wouldn't get generated). I have yet to make any measurements.

Regarding the 64-bit version, do you think it would also be worth the extra code to avoid incrementing them by 2 and instead specialize the return for each char in the word? e.g.:

```cs
DiffOffset0:

#if BIT64
    if (*a != *b) return *a - *b;
    if (*(a + 1) != *(b + 1)) return *(a + 1) - *(b + 1);
    if (*(a + 2) != *(b + 2)) return *(a + 2) - *(b + 2);
    if (*(a + 3) != *(b + 3)) return *(a + 3) - *(b + 3);
#else // ...
```